File: Python/15_manhattan_distance.py
import sys

from collections import namedtuple

Point = namedtuple('Point', ['x', 'y', 'ix'])
Value = namedtuple('Value', ['value', 'ix'])


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def read_points(reader, n):
    points = []
    for ix in range(n):
        x, y = convert_to_intlist(next(reader))
        points.append(Point(x, y, ix + 1))
    return points


def max_manhattan_distances(points):
    p0 = points[0]
    max_sum = Value(p0.x + p0.y, p0.ix)
    max_diff = Value(p0.x - p0.y, p0.ix)
    min_sum = Value(p0.x + p0.y, p0.ix)
    min_diff = Value(p0.x - p0.y, p0.ix)

    result = [(p0.ix, p0.ix)]
    for p in points[1:]:
        if p.x + p.y > max_sum.value:
            max_sum = Value(p.x + p.y, p.ix)
        if p.x + p.y < min_sum.value:
            min_sum = Value(p.x + p.y, p.ix)
        if p.x - p.y > max_diff.value:
            max_diff = Value(p.x - p.y, p.ix)
        if p.x - p.y < min_diff.value:
            min_diff = Value(p.x - p.y, p.ix)

        if (max_sum.value - min_sum.value > max_diff.value - min_diff.value):
            result.append((max_sum.ix, min_sum.ix))
        else:
            result.append((max_diff.ix, min_diff.ix))
    return result


def main():
    reader = sys.stdin
    n = int(next(reader))
    points = read_points(reader, n)
    result = max_manhattan_distances(points)
    for ix_pairs in result:
        print(' '.join([str(x) for x in ix_pairs]))


if __name__ == '__main__':
    main()

File: Haskell/15-Manhattan-distance.hs
import Control.Monad (replicateM)

data Point = Point { xCoord :: Int, yCoord :: Int, pointIx :: Int }
data Value = Value { value :: Int, index :: Int }


convertToIntList :: String -> [Int]
convertToIntList = map read . words


readPoints :: Int -> IO [Point]
readPoints n = do
    lines' <- replicateM n $ fmap convertToIntList getLine
    return $ map (\([x, y], ix) -> Point x y ix) $ zip lines' [1..]


maximalManhattanDistance :: [Point] -> [(Int, Int)]
maximalManhattanDistance points =
    let (Point x y ix) = head points
        go :: [Point] -> [(Int, Int)] -> Value -> Value -> Value -> Value -> [(Int, Int)]
        go [] acc _ _ _ _ = reverse acc
        go ((Point x y ix) : xss) acc maxSum maxDiff minSum minDiff =
            let maxSum' = if x + y > value maxSum then Value (x + y) ix else maxSum
                minSum' = if x + y < value minSum then Value (x + y) ix else minSum
                maxDiff' = if x - y > value maxDiff then Value (x - y) ix else maxDiff
                minDiff' = if x - y < value minDiff then Value (x - y) ix else minDiff
            in if (value maxSum' - value minSum' > value maxDiff' - value minDiff')
                then go xss ((index maxSum', index minSum') : acc) maxSum' maxDiff' minSum' minDiff'
                else go xss ((index maxDiff', index minDiff') : acc) maxSum' maxDiff' minSum' minDiff'
    in go (tail points) [(ix, ix)] (Value (x + y) ix) (Value (x - y) ix) (Value (x + y) ix) (Value (x - y) ix)


main :: IO()
main = do
    n <- readLn
    points <- readPoints n
    let result = maximalManhattanDistance points
    mapM_ (\(i, j) -> putStrLn $ unwords $ map show [i, j]) result
